plSqlStandards.md

```diff
@@ -0,0 +1,316 @@
+
+# Formatting Code
+
+## Important Recommendation: Do This Automatically
+
+Well formatted code greatly facilitates readability, maintainability and
+reliability. However, formatting code is also a chore: difficult and
+tedious to remember, time consuming to manually apply and difficult to
+police. Many code formatting rules, taken individually rather than as an
+entirety, will not individually add sufficiently to code readability,
+maintainability and reliability for their adoption to offset that cost.
+
+The way out of this to dilemma is to adopt tools that automatically
+format code. Indeed, unless the benefits of adopting any individual code
+formatting rule are very significant, or the costs and risks of not
+adopting such a rule are equally significant, a reasonable rule of thumb
+is that the only code formatting rules that should be adopted are those
+that can be applied automatically.
+
+In this spirit, all the code formatting rules described in this section
+are those that can be applied automatically in TOAD. See “APPENDIX: An
+Example of Automatically Formatting PL/SQL Code Using TOAD (the “Tool
+for Oracle Application Developers”)” for an example of how to do this.
+
+## Layout
+
+### Indents should be 3 characters wide. All code within loops and if statements should be so indented with one such indent for each level of nesting within if statements or loops.
+
+### THEN should be positioned on the next line, in-line with the related IF.
+
+### LOOP should be positioned on the next line, in-line with the associated DO or WHILE.
+
+### AS or IS should be positioned on the next line, in-line with the associated statement (such as with an associated CREATE).
+
+### Layout examples:
+
+BEGIN
+
+a := 1;
+
+d := 1;
+
+b := 2;
+
+WHILE TRUE
+
+LOOP
+
+a := 1;
+
+b := 2;
+
+IF a \> b
+
+THEN
+
+c := d;
+
+END IF;
+
+END LOOP;
+
+x := 1;
+
+c := d;
+
+END;
+
+CREATE PROCEDURE myproc
+
+AS
+
+BEGIN
+
+NULL;
+
+END;
+
+## Case and Lists
+
+### Oracle key words should be in uppercase.
+
+For example:
+
+SELECT aaa
+
+, bbb
+
+, cccc
+
+, SIN (x)
+
+FROM mytab;
+
+### Where you have lists, e.g. a list of column names or values, place each item on its own line, with commas, where needed, placed at the start of the line with a following space.
+
+This makes it far easier to make changes, such as inserting new items
+into the list, without losing track of the commas. For example:
+
+CREATE TABLE my\_table1
+
+(
+
+my\_num1 NUMBER
+
+, my\_num2 NUMBER
+
+);
+
+INSERT INTO building\_blocks (aaaaaaaaaaaaaaaa
+
+, bbbbbbbbbbbbbb
+
+, cccccccccccccccc
+
+, ddddddddddddddddd
+
+, eeeeeeeeeeeee)
+
+SELECT aaaaaaaaaaaaaaaa
+
+, bbbbbbbbbbbbbbbbbbb
+
+, cccccccccccccccccc
+
+, dddddddddddddddd
+
+, eeeeeeeeeeeeeee
+
+, fffffffffffffffff
+
+FROM mytab, histab, hertab
+
+WHERE histab.col1 = hertab.col2;
+
+# APPENDIX: An Example of Automatically Formatting PL/SQL Code Using TOAD (the “Tool for Oracle Application Developers”)
```

Comment:
I'd split this section out into a separate guidance document in order to fit with the structure of the repo.  I expect if we ask nicely @Cruikshanks would do this as part of rebasing to fit the document into the structure.

Reply:
Happy to do this. Would you like me to do this before hand so you can see what it would look like before it gets merged @nigejohnson ?

Reply:
> Happy to do this. Would you like me to do this before hand so you can see what it would look like before it gets merged @nigejohnson ?

OK Alan, thanks.

Reply:
I've reformatted the code blocks in the markdown now.
But I'm not clear what the plan is for making the "versioning" bit more general or splitting out the appendix into "guidance"? (I think those were the two outstanding things).